wagtail/snippets/tests/test_snippets.py

```diff
@@ -970,6 +1086,334 @@ def test_edit_snippet_with_revision(self):
+        )
+
+        # The instance should be updated
+        self.assertEqual(self.test_snippet.text, "Draft-enabled Bar, Now Published")
+
+        # The instance should be live
+        self.assertIs(self.test_snippet.live, True)
+        self.assertIs(self.test_snippet.has_unpublished_changes, False)
+        self.assertEqual(self.test_snippet.first_published_at, publish_timestamp)
+        self.assertEqual(self.test_snippet.last_published_at, publish_timestamp)
+        self.assertEqual(self.test_snippet.live_revision, latest_revision)
+
+        # The revision should be created and set as latest_revision
+        self.assertEqual(revisions.count(), 2)
+        self.assertEqual(latest_revision, revisions.last())
+
+        # The revision content should contain the new data
+        self.assertEqual(
+            latest_revision.content["text"],
+            "Draft-enabled Bar, Now Published",
+        )
+
+    def test_publish_then_save_draft(self):
+        publish_timestamp = now()
+        with freeze_time(publish_timestamp):
+            response = self.post(
+                post_data={
+                    "text": "Draft-enabled Bar, Published",
+                    "action-publish": "action-publish",
+                }
+            )
+
+        save_timestamp = now()
+        with freeze_time(save_timestamp):
+            response = self.post(
+                post_data={"text": "Draft-enabled Bar, Published and In Draft"}
+            )
+
+        self.test_snippet.refresh_from_db()
+        revisions = Revision.objects.for_instance(self.test_snippet).order_by("pk")
+        latest_revision = self.test_snippet.latest_revision
+
+        self.assertRedirects(
+            response, reverse("wagtailsnippets_tests_draftstatemodel:list")
+        )
+
+        # The instance should be updated with the last published changes
+        self.assertEqual(self.test_snippet.text, "Draft-enabled Bar, Published")
+
+        # The instance should be live
+        self.assertIs(self.test_snippet.live, True)
+        # The instance should have unpublished changes
+        self.assertIs(self.test_snippet.has_unpublished_changes, True)
+
+        self.assertEqual(self.test_snippet.first_published_at, publish_timestamp)
+        self.assertEqual(self.test_snippet.last_published_at, publish_timestamp)
+
+        # The live revision should be the first revision
+        self.assertEqual(self.test_snippet.live_revision, revisions.first())
+
+        # The second revision should be created and set as latest_revision
+        self.assertEqual(revisions.count(), 2)
+        self.assertEqual(latest_revision, revisions.last())
+
+        # The revision content should contain the new data
+        self.assertEqual(
+            latest_revision.content["text"],
+            "Draft-enabled Bar, Published and In Draft",
+        )
+
+    def test_publish_twice(self):
+        first_timestamp = now()
+        with freeze_time(first_timestamp):
```

Comment:
Ideally, we should only call the view once per test. Could you update this (and the above tests) to create the instance and set the `first_published_at` field manually?